### PR TITLE
PMPV0

### PR DESCRIPTION
**Pull request by ryley-o**
## Description of the change

Add PMPV0 contract per internal design document specs.

This PR implements the PMPV0 contract, as well as IWeb3Call interfaces.

⚠️ This PR does not integrate the PMP contract with the ONCHAIN system; that functionality will be added in a separate follow-on PR.

## Design

An abstract `Web3Call.sol` contract is included for any future iterations to use to ensure conformance to the web3call spec.

## Tests

Testing of all meaningful functionality and edge cases is included via hardhat tests.

### CODE REVIEW

File: packages/contracts/contracts/web3call/PMPV0.sol [hunk 1]
```diff
@@ -0,0 +1,979 @@
...
+        IPMPAugmentHook tokenPMPReadAugmentationHook
+    ) external {
+        // only artists may configure project hooks
+        _onlyArtist({
+            coreContract: coreContract,
+            projectId: projectId,
+            sender: msg.sender
+        });
+        // validation - check for ERC165 implementation for non-null hooks
+        if (address(tokenPMPPostConfigHook) != address(0)) {
```
Comment (lindsgil) [inline packages/contracts/contracts/web3call/PMPV0.sol:92]: What would this `tokenPMPPostConfigHook` be used for?

File: packages/contracts/contracts/web3call/PMPV0.sol [hunk 2]
```diff
@@ -0,0 +1,979 @@
...
+                revert("PMP: invalid authOption");
+            }
+        }
+
+        // ensure properly configured value
+        ParamType paramType = pmpConfigStorage.paramType;
+        // range checks for non-string params
+        if (paramType != ParamType.String) {
+            if (paramType == ParamType.Select) {
+                require(
```
Comment (lindsgil) [inline packages/contracts/contracts/web3call/PMPV0.sol:693]: Is there any validation we should do on the individual select values?

File: packages/contracts/contracts/web3call/PMPV0.sol (hunk 1 above)
Comment (ryley-o) [reply packages/contracts/contracts/web3call/PMPV0.sol:92]: it can be used for effects of PMP updates.
one example:
* a PMP named "color" is set to "red"
* a post-config hook tallies all PMP edits, and reverts if the total number of "red" is above some number, across the entire project

another example:
* a post-config hook tallies the total of each option in a 3-option "select" type PMP
* an augmentation hook grabs this tallied value and injects it into the returned PMPs for the token
* the project's script in the generator now has the total PMP tallies for an entire project, in addition to the individual configuration for the token

its very open ended and could get pretty wild 😆

**Review by yoshiwarab — APPROVED** [review]

File: packages/contracts/contracts/web3call/PMPV0.sol (hunk 1 above)
Comment (lindsgil) [reply packages/contracts/contracts/web3call/PMPV0.sol:92]: makes sense, thanks for the context!

File: packages/contracts/contracts/web3call/PMPV0.sol (hunk 2 above)
Comment (lindsgil) [reply packages/contracts/contracts/web3call/PMPV0.sol:693]: chatted through this offline, going to resolve~!

File: packages/contracts/contracts/web3call/PMPV0.sol [hunk 3]
```diff
@@ -0,0 +1,979 @@
...
+            // update highestConfigNonce
+            pmpConfigStorage.highestConfigNonce = newConfigNonce;
+            // copy function input pmpConfig data to pmpConfigStorage
+            PMPConfig memory inputPMPConfig = pmpInputConfigs[i].pmpConfig;
+            pmpConfigStorage.authOption = inputPMPConfig.authOption;
+            pmpConfigStorage.paramType = inputPMPConfig.paramType;
+            pmpConfigStorage.pmpLockedAfterTimestamp = inputPMPConfig
+                .pmpLockedAfterTimestamp;
+            pmpConfigStorage.authAddress = inputPMPConfig.authAddress;
+            // TODO - use immutableStringArray for selectOptions
```
[the remark this answers is not in the document]
Comment (ryley-o) [reply packages/contracts/contracts/web3call/PMPV0.sol:200]: ah, I meant to implement this here, will update!

File: packages/contracts/contracts/web3call/PMPV0.sol [hunk 4]
```diff
@@ -0,0 +1,979 @@
...
+
+    bytes32 private constant _TYPE = "PMPV0";
+
+    bytes16 private constant _HEX_DIGITS = "0123456789abcdef";
+    uint256 private constant _DECIMAL_PRECISION_DIGITS = 10;
+    uint256 private constant _DECIMAL_PRECISION =
+        10 ** _DECIMAL_PRECISION_DIGITS;
+
+    // @dev min hex color assumed to be 0x000000
+    uint256 private constant _HEX_COLOR_MAX = 0xFFFFFF;
```
[the remark this answers is not in the document]
Comment (ryley-o) [reply packages/contracts/contracts/web3call/PMPV0.sol:46]: as discussed, leaving alpha channel as optionally defined in a separate PMP.
added dev comment to clarify in [943ff41](https://github.com/ArtBlocks/artblocks-contracts/pull/1764/commits/943ff41b712d665cba9194bdcb294e81e3d3d9a4)

File: packages/contracts/contracts/web3call/PMPV0.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryley-o) [reply packages/contracts/contracts/web3call/PMPV0.sol:201]: actually, going to do this in a follow-on PR to decouple all the refactor 🙏
